client/go/outline/vpn/nmconn_linux.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vpn
+
+import (
+	"encoding/binary"
+	"log/slog"
+	"net"
+	"time"
+
+	gonm "github.com/Wifx/gonetworkmanager/v2"
+	"golang.org/x/sys/unix"
+)
+
+type nmConnectionOptions struct {
+	Name            string
+	TUNName         string
+	TUNAddr4        net.IP
+	DNSServers4     []net.IP
+	FWMark          uint32
+	RoutingTable    uint32
+	RoutingPriority uint32
+}
+
+func (c *linuxVPNConn) establishNMConnection() (err error) {
+	defer func() {
+		if err != nil {
+			c.closeNMConnection()
+		}
+	}()
+
+	if c.nm, err = gonm.NewNetworkManager(); err != nil {
```

Comment:
Can you set `c.nm` and `c.ac` in the caller instead? You will need to return it there.

It's hard to track the state in the linuxVPNConn code. That will make the side-effect a bit clearer. It's hard to track the mutations when they are hidden as side-effects.

Reply:
Also, it would be helpful if the methods were where the struct is defined. Why are they separate?